src/allocator_defrag.c

```diff
@@ -0,0 +1,379 @@
+ * This interface (named MIB in jemalloc) is faster as it avoids string dict lookup at run-time. */
+static inline int jeQueryCtlInterface(const jeMallctlKey *key_info, void *value, size_t *size) {
+    return je_mallctlbymib(key_info->key, key_info->keylen, value, size, NULL, 0);
+}
+
+static int binQueryHelperInitialization(jeBinInfoKeys *helper, unsigned bin_index) {
+    char buf[128];
+
+    /* Mib of fetch number of used regions in the bin */
+    snprintf(buf, sizeof(buf), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.curregs", bin_index);
+    if (jeQueryKeyInit(buf, &helper->curr_regs) != 0) return -1;
+    /* Mib of fetch number of current slabs in the bin */
+    snprintf(buf, sizeof(buf), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.curslabs", bin_index);
+    if (jeQueryKeyInit(buf, &helper->curr_slabs) != 0) return -1;
+    /* Mib of fetch nonfull slabs */
+    snprintf(buf, sizeof(buf), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.nonfull_slabs", bin_index);
+    if (jeQueryKeyInit(buf, &helper->nonfull_slabs) != 0) return -1;
+
+    return 0;
+}
+
+/* Initializes the defragmentation system for the jemalloc memory allocator.
+ *
+ * This function performs the necessary setup and initialization steps for the defragmentation system.
+ * It retrieves the configuration information for the jemalloc arenas and bins, and initializes the usage
+ * statistics data structure.
+ *
+ * return 0 on success, or a non-zero error code on failure.
+ *
+ * The initialization process involves the following steps:
+ * 1. Check if defragmentation is supported by the current jemalloc version.
+ * 2. Retrieve the arena bin configuration information using the `je_mallctlbymib` function.
+ * 3. Initialize the `usage_latest` structure with the bin usage statistics and configuration data.
+ * 4. Set the `defrag_supported` flag to indicate that defragmentation is enabled.
+ *
+ * Note: This function must be called before using any other defragmentation-related functionality.
+ * It should be called during the initialization phase of the code that uses the
+ * defragmentation feature.
+ */
+int allocatorDefragInit(void) {
+    char buf[100];
+    jeBinInfo *bin_info;
+    unsigned nbins;
+    size_t sz = sizeof(nbins);
+    int je_res = 0;
+
+    // the init should be called only once, fail if unexpected call
+    assert(!defrag_supported);
+
+    // Get the mib of the per memory pointers query command that is used during defrag scan over memory
+    if (jeQueryKeyInit("experimental.utilization.batch_query", &arena_bin_conf.util_batch_query) != 0) return -1;
+
+    je_res = jeQueryKeyInit("epoch", &arena_bin_conf.epoch);
+    assert(je_res == 0);
+    jeRefreshStats(&arena_bin_conf);
+
+    size_t len = sizeof(jemalloc_quantum);
+    je_mallctl("arenas.quantum", &jemalloc_quantum, &len, NULL, 0);
+    // lg-quantum should be 3
+    assert(jemalloc_quantum == 8);
```

Comment:
How much of this code depends on lg-quantum == 3? Just wondering because if we some day want to change to the default 4, how hard would that be?

Reply:
previous iteration of the code had support for quantum=4, removed it following @madolson review comment. It's not hard to re-add it at all, it only affects the bin id's when we need to get bin from region size.